madprops/serializers.py

```diff
@@ -61,95 +91,160 @@ class Property(model):
             return json.loads(obj.value)
         return obj.value
 
-    @cached_property
-    def errors(self):
-        if not isinstance(self.init_data, dict):
-            return {'non_field_errors': ['Expected a dictionary.']}
-
-        # Ensure we don't modify read-only properties.  Remove them from input.
-        for prop in self.opts.read_only_props:
-            self.init_data.pop(prop, None)
-
-        result = RelationsList()
-        existent_props = dict((obj.name, obj) for obj in self.objects)
-
-        errors = {}
-        for name, value in self.init_data.iteritems():
-            self.object = existent_props.get(name)
-            updated = self.from_native({name: value})
-            errors.update(self._errors)
-            if not self._errors:
-                result.append(updated)
-        self.object = result
-        return errors
-
-    def from_native(self, data, files=None):
-        name, value = data.iteritems().next()
-        # Deal with JSON properties
-        if name in self.opts.json_props:
-            value = json.dumps(value)
-
-        data = {'name': name, 'value': value}
-        data = self._from_native_hook(data)
-        return super(PropertySerializer, self).from_native(data, files)
-
-    def _from_native_hook(self, data):
-        # Update created property with reference to parent object
-        parent_obj_field = self.opts.parent_obj_field
-        parent_id_field = parent_obj_field + '_id'
-        data[parent_obj_field] = self.context['view'].kwargs[parent_id_field]
-        return data
+    def save(self, property_data=None):
+        property_data = property_data or self.validated_data
+        prop_name = property_data['name']
 
-    @property
-    def objects(self):
-        # Ensure we always work with iterable
-        if isinstance(self.object, Iterable):
-            return self.object
-
-        return [] if self.object is None else [self.object]
-
-    def save_object(self, obj, **kwargs):
-        # Ensure we have only one property with the same name
-        model = self.opts.model
+        # Try to find property by it's name.
         parent_obj_field = self.opts.parent_obj_field
         filters = {
-            parent_obj_field: getattr(obj, parent_obj_field),
-            'name': obj.name
+            parent_obj_field: property_data[parent_obj_field],
+            'name': prop_name
         }
-        if not model.objects.filter(**filters).update(value=obj.value):
-            obj.save(**kwargs)
-
 
-class NestedPropertySerializer(PropertySerializer):
-    """Version of PropertySerializer for nested resources
-
-    Intended to be used as a base class for serializer for property resource
-    exposed as a nested resource.
-    """
+        # If it already exists - update it's value. Otherwise - create a new
+        # property.
+        prop = self.Meta.model.objects.filter(**filters).first()
+
+        if prop_name in self.opts.read_only_props:
+            return prop
```

Comment:
Maybe we should handle `self.opts.read_only_props` in `to_internal_value` rather than `save()`?